packages/@aws-cdk/aws-lambda-event-sources/lib/stream.ts

```diff
@@ -0,0 +1,52 @@
+import lambda = require('@aws-cdk/aws-lambda');
+import {Duration} from "@aws-cdk/core";
+
+export interface StreamingEventSourceProps {
+  /**
+   * The largest number of records that AWS Lambda will retrieve from your event
+   * source at the time of invoking your function. Your function receives an
+   * event with all the retrieved records.
+   *
+   * Valid Range: Minimum value of 1. Maximum value of 10000.
+   *
+   * @default 100
+   */
+  readonly batchSize?: number;
+
+  /**
+   * Where to begin consuming the stream.
+   */
+  readonly startingPosition: lambda.StartingPosition;
+
+  /**
+   * The maximum amount of time to gather records before invoking the function.
+   * Maximum of Duration.minutes(5)
+   *
+   * @default Duration.seconds(0)
+   */
+  readonly maxBatchingWindow?: Duration;
+}
+
+/**
+ * Use an stream as an event source for AWS Lambda.
+ */
+export abstract class StreamEventSource implements lambda.IEventSource {
+  protected constructor(protected readonly props: StreamingEventSourceProps, protected readonly maxBatchSize: number) {
+    if (this.props.batchSize !== undefined && (this.props.batchSize < 1 || this.props.batchSize > maxBatchSize)) {
+      throw new Error(`Maximum batch size must be between 1 and ${maxBatchSize} inclusive (given ${this.props.batchSize})`);
+    }
+  }
+
+  public bind(_target: lambda.IFunction): void {
+    throw new Error('Cannot bind StreamEventSource, use DynamoEventSource or KinesisEventSource');
+  }
+
+  protected getEventSourceMappingOptions(eventSourceArn: string): lambda.EventSourceMappingOptions {
```

Comment:
Same here about over-delegation. Don't pass the `eventSourceArn` here. Instead insert it into the struct, returned by this method, at the calling point in the subclasses.

Reply:
This would break the requirement of the `EventSourceMappingOptions` return type, which I'd like to keep. I could change it to `Omit<EventSourceMappingOptions, 'eventSourceArn'>`, I'll see if JSII is alright with it.

Reply:
I see, good point.

How about renaming the method signature to `enrichMappingOptions(options: lambda.EventSourceMappingOptions)` and either adding the stream-related properties and return it back?